Jetpack Pro Dashboard: Implement showing monitor settings unsaved changes warning alert.

yashwin opened the pull request:
Related to 1204408201748644-as-1204666762460281

## Proposed Changes

This PR implements showing monitor settings unsaved changes warning alert.

#### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Run `git checkout add/show-email-verified-badge-for-10sec` and `yarn start-jetpack-cloud`
2. Open http://jetpack.cloud.localhost:3000/, and you'll be redirected to the /dashboard.
3. Enable monitor if not enabled already.
4. Click on the clock icon next to the monitor toggle.
5. Make any change to duration, toggle mobile and email notifications, or add/edit/remove the email items. Click the `Cancel` or `X` button or `ESC` key and verify that an error message is as shown below. Clicking on the `Cancel` or `X` button or the `ESC` key should close the popup. Please test all the changes individually and verify the error is shown as expected. 

<img width="476" alt="Screenshot 2023-05-29 at 3 38 29 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/c217ab81-ca4d-4b76-a794-1ac16f358e38">

<img width="472" alt="Screenshot 2023-06-01 at 5 00 35 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/78922a49-aecd-4825-a0d1-a3ab87ea262f">


6. Open the Jetpack live link and perform the same steps and verify the error message is shown when only duration, toggle mobile/email is changed. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

yashwin committed:
Implement showing monitor settings unsaved changes warning.

yashwin committed:
Set role as alert for error messages

yashwin committed:
Check email only if feature is enabled

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/types.ts:
```diff
@@ -289,3 +289,12 @@ export interface StateMonitorSettingsEmail extends MonitorSettingsEmail {
 
 export type AllowedMonitorContactActions = 'add' | 'verify' | 'edit' | 'remove';
+
+export type MonitorDuration = { label: string; time: number };
```
👍

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx:
```diff
@@ -61,11 +61,45 @@ export default function NotificationSettings( {
 
+	// Check if any unsaved changes are present and prompt user to confirm before closing the modal.
+	const handleOnClose = useCallback( () => {
+		const unsavedChangesExist =
```
It's good that we are now tracking changes. Maybe we can move `unsavedChangesExist` outside this callback and use it to disable the 'Save' button at the same time if this is **false**. wdyt?

yashwin replied:
Agree. I have made the changes - https://github.com/Automattic/wp-calypso/pull/77485/commits/416321f7be7c9a159e078a1dddf2569973d2e703

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx:
```diff
@@ -61,11 +61,45 @@ export default function NotificationSettings( {
 	const [ isAddEmailModalOpen, setIsAddEmailModalOpen ] = useState< boolean >( false );
 	const [ selectedEmail, setSelectedEmail ] = useState< StateMonitorSettingsEmail | undefined >();
 	const [ selectedAction, setSelectedAction ] = useState< AllowedMonitorContactActions >();
+	const [ initialSettings, setInitialSettings ] = useState< InitialMonitorSettings >( {
```
I'm curious why we do not use the `settings` props for tracking changes?

yashwin replied:
We can. I tried doing this. But, it is tricky since we also have `bulkSettings` also. I would like to keep a different state for initial settings considering we do some pre-processing to the settings.

jkguidaven replied:
That makes sense. Thanks for explaining this. 👍

jkguidaven reviewed (commented):
Nice work on this @yashwin 👍 . I left some few comments.

Also I would suggest having design review on this.

cc. @ilonagl

yashwin committed:
Disable Save button if no changes exist

yashwin commented:
> Nice work on this @yashwin 👍 . I left some few comments.

Thanks for your review! I have addressed your comments.

> Also I would suggest having design review on this.

I have pinged @ilonagl for her review.

jkguidaven reviewed (approved):
LGTM. thanks for addressing my comments.

yashwin committed:
Address PR comments

yashwin committed:
Disable save button only for single site update not bulk edit

yashwin commented:
Thanks for the review and feedback. I have addressed adding the dot at the end of the sentence and renaming the button to `Back` - https://github.com/Automattic/wp-calypso/pull/77485/commits/72cae35e2a3b1b3cf2feddc69c6fb6a35dfa2233. Since the resend is being implemented in another PR, I'll make sure the make the changes there.

yashwin committed:
Set initial settings for bulk update

ilonagl commented:
Thank you, @yashwin! Much appreciated.
